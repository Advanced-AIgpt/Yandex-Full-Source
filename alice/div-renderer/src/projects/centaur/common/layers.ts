// Device model stack layers
//  https://a.yandex-team.ru/arc/trunk/arcadia/smart_devices/android/centaur-app/src/main/java/ru/yandex/quasar/centaur_app/stack/contract/Layer.kt
export enum Layer {
    ROOT,
    CONTENT,
    DIALOG,
    SHUTTER,
    ALARM,
}
