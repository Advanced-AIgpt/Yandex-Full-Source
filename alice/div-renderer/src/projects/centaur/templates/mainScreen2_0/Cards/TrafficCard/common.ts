import { ITrafficForecast } from './types';

export function sortTimeOfDay(originalArray: Readonly<ITrafficForecast[]>): ITrafficForecast[] {
    const forecastsFirst = [];
    const forecastsLast = [];

    for (const forecastElement of originalArray) {
        if (forecastElement.time === null || typeof forecastElement.time === 'undefined') {
            continue;
        }

        if (forecastsFirst.length === 0) {
            forecastsFirst.push(forecastElement);
        } else if (forecastsLast.length > 0) {
            forecastsLast.push(forecastElement);
        } else if (forecastElement.time - (forecastsFirst[forecastsFirst.length - 1].time as number) === 1) {
            forecastsFirst.push(forecastElement);
        } else {
            forecastsLast.push(forecastElement);
        }
    }

    return [...forecastsLast, ...forecastsFirst];
}
