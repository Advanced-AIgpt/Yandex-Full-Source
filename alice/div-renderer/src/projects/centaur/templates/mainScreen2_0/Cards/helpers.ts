export function getCardMainScreenId({ rowIndex, colIndex }: { rowIndex?: number; colIndex?: number }) {
    return `main_screen_card_${rowIndex || 0}_${colIndex || 0}`;
}
