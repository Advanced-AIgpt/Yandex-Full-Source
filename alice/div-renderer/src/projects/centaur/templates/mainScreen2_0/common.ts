import { IDivAction } from 'divcard2';
import { EnumLayer } from '../../actions/client';
import { CrutchLoader } from '../../components/Loader/Loader';

interface IPutLoaderToActionOptions {
    action?: string | null,
    layer: EnumLayer,
    actionLogId?: string,
}

export const putLoaderToAction = ({
    action,
    layer,
    actionLogId = 'action_with_loader',
}: IPutLoaderToActionOptions): IDivAction[] | undefined => {
    if (!action) {
        return undefined;
    }
    return [
        CrutchLoader(layer),
        {
            log_id: actionLogId,
            url: action,
        },
    ];
};
