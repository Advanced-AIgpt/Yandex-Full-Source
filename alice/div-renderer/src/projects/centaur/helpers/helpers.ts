/* eslint-disable camelcase */
import { sortedIndexBy } from 'lodash';
import { format, isSameDay } from 'date-fns';
import { ru } from 'date-fns/locale';
import { Div, IDivData, TextBlock } from 'divcard2';
import { IRequestState } from '../../../common/types/common';
import { getTemplates, templatesRegistry } from '../../../registries/common';
import { centaurTemplatesClass } from '../index';
import { Palette } from './palette';

export { default as pluralize } from '@yandex-int/i18n/lib/plural/ru';
type TextProps = Pick<ConstructorParameters<typeof TextBlock>[0], 'text_color'|'font_size'|'font_weight'|'line_height'>;

export const textBreakpoint =
    (...breakpoints: [number, TextProps][]) =>
        (text: string) => {
            const idx = Math.min(sortedIndexBy(breakpoints, [text.length, {}], 0), breakpoints.length - 1);
            return { text, ...breakpoints[idx][1] };
        };

export const formatDate = (d: Date | string | number, f: string) => format(new Date(d), f, { locale: ru });

export const textCommand = (c: string) => `dialog://text_command?query=${encodeURI(c)}`;

export const updatedAt = (d: Date) => {
    const now = new Date();
    return `Обновлен ${isSameDay(now, d) ? 'сегодня' : formatDate(d, 'dd.MM')}`;
};

export const insertBetween = <I extends unknown, S extends unknown>(itemList: I[], separator: S) => {
    return itemList.reduce<(I | S)[]>((result, item, index) => {
        if (index !== 0) {
            result.push(separator);
        }

        result.push(item);

        return result;
    }, []);
};

export interface ITemplateCard {
    templates: {
        [name: string]: Div;
    };
    palette?: Palette;
    card: IDivData;
}

// Дива верхнего уровня. Замена для new `TemplateCard()`
export const TopLevelCard = (card: Omit<IDivData, 'variable_triggers' | 'variables'>, requestState: IRequestState, palette?: Palette) => {
    const templates = Object.assign(
        getTemplates(requestState.localTemplates, templatesRegistry),
        centaurTemplatesClass.getUsedTemplates(requestState),
        requestState.res.localTemplates.getAll(),
    );

    return {
        templates,
        palette,
        card: {
            ...card,
            variable_triggers: requestState.variableTriggers.getAll(),
            variables: [...requestState.variables],
        } as IDivData,
    };
};
