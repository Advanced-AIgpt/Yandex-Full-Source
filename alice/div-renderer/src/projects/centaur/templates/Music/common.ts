import {
    ContainerBlock,
    Div,
    FixedSize,
    IDivAction,
    ImageBackground,
    ImageBlock, MatchParentSize,
    SolidBackground,
    TextBlock,
} from 'divcard2';
import { colorBlackOpacity70, offsetFromEdgeOfScreen } from '../../style/constants';
import { compact } from 'lodash';
import EmptyDiv from '../../components/EmptyDiv';
import { text28m } from '../../style/Text/Text';
import { directivesAction } from '../../../../common/actions';
import { centaurMusicPlay } from '../../../../common/actions/server/musicActions';
import MusicPlaying from './MusicPlaying/MusicPlaying';
import { IRequestState } from '../../../../common/types/common';
import LikeButton from './components/LikeButton';

interface IGetPlaylistItemProps {
    title: string;
    subtitle?: string;
    objectId: string;
    objectType: string;
    startFromTrackId?: string;
    isCurrent?: boolean;
    imageUrl: string;
    imageBackground?: string;
    requestState: IRequestState;
    isLiked?: boolean;
}

function CoverWithBackground(imageUrl: string, background: string, action: IDivAction, isCurrent: boolean, requestState: IRequestState) {
    return new ContainerBlock({
        width: new FixedSize({ value: 80 }),
        height: new FixedSize({ value: 80 }),
        alignment_vertical: 'top',
        actions: [action],
        margins: {
            right: 24,
        },
        border: {
            corner_radius: 40,
        },
        content_alignment_horizontal: 'center',
        content_alignment_vertical: 'center',
        background: compact([
            new SolidBackground({
                color: background,
            }),
        ]),
        orientation: 'overlap',
        items: compact([
            new ImageBlock({
                image_url: imageUrl,
                width: new FixedSize({ value: 48 }),
                height: new FixedSize({ value: 48 }),
            }),
            isCurrent && new EmptyDiv({
                width: new FixedSize({ value: 80 }),
                height: new FixedSize({ value: 80 }),
                background: [
                    new SolidBackground({
                        color: colorBlackOpacity70,
                    }),
                ],
            }),
            isCurrent && MusicPlaying(requestState),
        ]),
    });
}

function CoverWithoutBackground(imageUrl: string, action: IDivAction, isCurrent: boolean, requestState: IRequestState) {
    return new ContainerBlock({
        width: new FixedSize({ value: 80 }),
        height: new FixedSize({ value: 80 }),
        alignment_vertical: 'top',
        content_alignment_horizontal: 'center',
        content_alignment_vertical: 'center',
        actions: [action],
        margins: {
            right: 24,
        },
        border: {
            corner_radius: 20,
        },
        background: compact([
            new ImageBackground({
                image_url: imageUrl,
            }),
            isCurrent && new SolidBackground({
                color: colorBlackOpacity70,
            }),
        ]),
        items: [
            isCurrent ? MusicPlaying(requestState) : new EmptyDiv(),
        ],
    });
}

export function getPlaylistItem({
    isCurrent = false,
    objectType,
    subtitle,
    objectId,
    title,
    startFromTrackId,
    imageUrl,
    imageBackground,
    requestState,
    isLiked,
}: IGetPlaylistItemProps): {
    options: Partial<ConstructorParameters<typeof ContainerBlock>[0]>;
    items: Div[];
} {
    const openAction = {
        log_id: 'play_music_click',
        url: directivesAction(centaurMusicPlay({
            disableNlg: true,
            objectType,
            objectId,
            startFromTrackId,
        })),
    };

    return {
        options: {
            orientation: 'horizontal',
            paddings: {
                top: 28,
                left: offsetFromEdgeOfScreen,
                bottom: 28,
                right: offsetFromEdgeOfScreen,
            },
            content_alignment_vertical: 'center',
        },
        items: compact([
            imageBackground ?
                CoverWithBackground(imageUrl, imageBackground, openAction, isCurrent, requestState) :
                CoverWithoutBackground(imageUrl, openAction, isCurrent, requestState),
            new ContainerBlock({
                width: new MatchParentSize({ weight: 1 }),
                actions: [openAction],
                items: compact([
                    new TextBlock({
                        ...text28m,
                        text: title,
                    }),
                    subtitle && new TextBlock({
                        ...text28m,
                        text: subtitle,
                        alpha: 0.5,
                    }),
                ]),
            }),
            typeof isLiked !== 'undefined' && LikeButton({ isLiked, trackId: startFromTrackId || '' }),
        ]),
    };
}
