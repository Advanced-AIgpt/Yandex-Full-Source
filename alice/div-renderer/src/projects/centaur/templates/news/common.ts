import {
    ContainerBlock,
    DivBackground,
    FixedSize, IDivAction,
    ImageBackground,
    ImageBlock,
    MatchParentSize,
    SolidBackground,
    TextBlock,
    WrapContentSize,
} from 'divcard2';
import { compact } from 'lodash';
import { utcToZonedTime } from 'date-fns-tz';
import { differenceInCalendarDays, isSameDay } from 'date-fns';
import { Long } from 'protobufjs';
import { NAlice } from '../../../../protos';
import { formatDate } from '../../helpers/helpers';
import {
    colorDarkGrey1Opacity70,
    colorReadButton,
    colorWhite,
    colorWhiteOpacity60,
    offsetFromEdgeOfScreen,
} from '../../style/constants';
import { text40r, title32m, title60m } from '../../style/Text/Text';
import { getImageByTopic } from '../../helpers/newsHelper';
import { Avatar } from '../../../../common/helpers/avatar';

export const patchToOrig = (avatarsUrl: string) => avatarsUrl.replace(/\/[^/]+$/, '/orig');

export type NewsCardProps = {
    item: NAlice.NData.ITNewsItem;
    topic: string;
    bottomGap: boolean;
    tz: string;
};

export function newsCommonCard(
    { item: { Text, Image, Agency, Logo, PubDate, ExtendedNews }, topic, bottomGap, tz }: NewsCardProps,
    openExtendedNewsDivAction: IDivAction) {
    return new ContainerBlock({
        width: new MatchParentSize(),
        height: new MatchParentSize(),
        background: generateBackground(topic, Image),
        content_alignment_horizontal: 'left',
        content_alignment_vertical: 'bottom',
        paddings: {
            left: offsetFromEdgeOfScreen,
            bottom: bottomGap ? 104 : offsetFromEdgeOfScreen,
            right: offsetFromEdgeOfScreen,
        },
        items: [
            new ContainerBlock({
                width: new WrapContentSize(),
                height: new MatchParentSize({ weight: 1 }),
                content_alignment_vertical: 'bottom',
                orientation: 'horizontal',
                margins: {
                    bottom: 20,
                },
                items: newsSource(Agency, Logo, PubDate, tz),
            }),
            new ContainerBlock({
                width: new MatchParentSize(),
                height: new WrapContentSize(),
                content_alignment_vertical: 'bottom',
                orientation: 'horizontal',
                items: compact([
                    new TextBlock({
                        ...title60m,
                        width: new MatchParentSize({ weight: 1 }),
                        height: new WrapContentSize(),
                        max_lines: 5,
                        truncate: 'end',
                        text: Text ?? ' ',
                    }),
                    ExtendedNews?.length && new TextBlock({
                        ...text40r,
                        width: new WrapContentSize(),
                        height: new WrapContentSize(),
                        background: [new SolidBackground({
                            color: colorReadButton,
                        })],
                        border: {
                            corner_radius: 24,
                        },
                        paddings: {
                            top: 22,
                            right: 33,
                            bottom: 22,
                            left: 33,
                        },
                        margins: {
                            left: 48,
                        },
                        alignment_horizontal: 'right',
                        text: 'Читать',
                        action: openExtendedNewsDivAction,
                    }),
                ]),
            }),
        ],
    });
}

export const newsSource = (
    Agency: (string | null | undefined),
    Logo: (string | null | undefined),
    PubDate: (number | Long | null | undefined),
    tz: string,
) =>
    compact([
        Logo && new ContainerBlock({
            background: [
                new SolidBackground({
                    color: colorWhite,
                }),
            ],
            content_alignment_horizontal: 'center',
            content_alignment_vertical: 'center',
            border: {
                corner_radius: 12,
            },
            width: new FixedSize({ value: 40 }),
            height: new FixedSize({ value: 40 }),
            margins: {
                right: 20,
            },
            items: [
                new ImageBlock({
                    width: new FixedSize({ value: 28 }),
                    height: new FixedSize({ value: 28 }),
                    image_url: Logo ?? '',
                }),
            ],
        }),
        new TextBlock({
            ...title32m,
            text_color: colorWhiteOpacity60,
            width: new WrapContentSize(),
            height: new WrapContentSize(),
            text_alignment_vertical: 'center',
            text: (Agency ? Agency + ' • ' : '') + (Number(PubDate) ? formDate(Number(PubDate) * 1000 || 0, tz) : ''),
        }),
    ]);

export const generateBackground = (
    Topic: string | null | undefined,
    NewsImage: NAlice.NData.TNewsItem.ITImage | null | undefined,
    color = colorDarkGrey1Opacity70,
) => {
    const result: DivBackground[] = [];
    if (NewsImage && NewsImage.Src) {
        const avatar = Avatar.fromUrl(NewsImage.Src);

        result.push(
            new ImageBackground({
                image_url: avatar ? avatar.setTypeName('orig').toString() : NewsImage.Src,
                scale: 'fill',
                preload_required: 1,
            }),
        );
    } else {
        result.push(
            new ImageBackground({
                image_url: getImageByTopic(Topic),
                preload_required: 1,
            }),
        );
    }

    result.push(new SolidBackground({ color }));
    return result;
};

export const formDate = (date: number, tz: string) => {
    const d = utcToZonedTime(date, tz);
    const now = utcToZonedTime(Date.now(), tz);
    if (isSameDay(d, now)) {
        return formatDate(d, 'HH:mm');
    }
    if (differenceInCalendarDays(now, d) === 1) {
        return `вчера в ${formatDate(d, 'HH:mm')}`;
    }
    return formatDate(d, 'd MMMM в HH:mm');
};
