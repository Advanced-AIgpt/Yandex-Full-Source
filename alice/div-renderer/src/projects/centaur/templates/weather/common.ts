import {
    ContainerBlock,
    Div, FixedSize,
    GalleryBlock, ImageBackground,
    ImageBlock, MatchParentSize, SolidBackground,
    Template,
    templateHelper,
    TextBlock,
    WrapContentSize,
} from 'divcard2';
import { capitalize } from 'lodash';
import { differenceInCalendarDays } from 'date-fns';
import { NAlice } from '../../../../protos';
import { Void } from '../../../../common/types/util';
import { formatDate, pluralize } from '../../helpers/helpers';
import { text40m, title146m, title48m, title56m, title64m, title90m } from '../../style/Text/Text';
import { colorWhiteOpacity40, colorWhiteOpacity50, colorWhiteOpacity53 } from '../../style/constants';
import { getS3Asset } from '../../helpers/assets';

const s3Url = (part: string) => getS3Asset(`weather/${part}`);
export const getWeatherImage = (daypart: string, condition: string) => s3Url('fact_bg_' + daypart + '_' + condition) + '.png';

const cloudiness = (cloudiness: number, precStrength: number) => {
    if (cloudiness < 0.01 && precStrength < 0.01) {
        return 'clear';
    }

    if (precStrength < 0.01) {
        if (cloudiness > 0.75) {
            return 'overcast_light_prec';
        }
        return 'cloudy';
    }

    if (precStrength < 0.5) {
        return 'overcast_light_prec';
    }

    return 'overcast_prec';
};

export const dayPartsForms: Record<string, string> = {
    morning: 'утром',
    day: 'днем',
    evening: 'вечером',
    night: 'ночью',
};
export const formatTemp = (temp: number) => `${temp > 0 ? '+' : ''}${temp}°`;

export const formatHour = (hour: number) => `${hour < 10 ? '0' : ''}${hour}:00`;

export const calcDaypart = (sunrise: Void<string>, sunset: Void<string>, userTime: Void<string>) =>
    sunrise && sunset && userTime && userTime > sunrise && userTime < sunset ? 'day' : 'night';

export const calcCondition = (cloudness: number, precStrength: number) => cloudiness(cloudness, precStrength);

export const formWeatherIconUrl = (iconType: string, size: number) => s3Url(`${iconType}_${size}.png`);

type CalcWhenProps = {
    userDate: Date;
    date: Date;
    daypart: string | undefined;
};
export const calcWhen = ({ userDate, date, daypart }: CalcWhenProps) => {
    const result: string[] = [];
    result.push(['Сейчас', 'Завтра', 'Послезавтра'][differenceInCalendarDays(userDate, date)]);
    if (daypart) {
        result.push(dayPartsForms[daypart]);
    }
    return result.join(' ');
};

export const weatherTemplates = {
    weatherCard: new ContainerBlock({
        width: new FixedSize({ value: 150 }),
        height: new FixedSize({ value: 300 }),
        orientation: 'vertical',
        content_alignment_horizontal: 'center',
        items: [
            new TextBlock({
                ...text40m,
                width: new WrapContentSize(),
                margins: {
                    top: 24,
                },
                text: new Template('time'),
            }),
            new ImageBlock({
                width: new FixedSize({ value: 100 }),
                height: new FixedSize({ value: 100 }),
                margins: {
                    top: 20,
                },
                image_url: new Template('icon'),
            }),
            new TextBlock({
                ...title48m,
                width: new WrapContentSize(),
                margins: {
                    top: 20,
                },
                text: new Template('temp'),
            }),
        ],
    }),
    weatherDayCard: new ContainerBlock({
        orientation: 'vertical',
        width: new FixedSize({ value: 278 }),
        height: new WrapContentSize(),
        content_alignment_horizontal: 'center',
        items: [
            new TextBlock({
                ...text40m,
                text: new Template('weekDay'),
                width: new WrapContentSize(),
            }),
            new TextBlock({
                ...text40m,
                text_color: colorWhiteOpacity40,
                text: new Template('date'),
                width: new WrapContentSize(),
            }),
            new ImageBlock({
                image_url: new Template('icon'),
                width: new FixedSize({ value: 148 }),
                height: new FixedSize({ value: 148 }),
                margins: {
                    top: 32,
                },
            }),
            new ContainerBlock({
                orientation: 'horizontal',
                width: new WrapContentSize(),
                height: new WrapContentSize(),
                margins: {
                    top: 32,
                },
                content_alignment_vertical: 'bottom',
                items: [
                    new TextBlock({
                        ...title56m,
                        text: new Template('dayTemp'),
                        width: new WrapContentSize(),
                    }),
                    new TextBlock({
                        ...title56m,
                        text_color: colorWhiteOpacity50,
                        text: new Template('nightTemp'),
                        width: new WrapContentSize(),
                    }),
                ],
            }),
        ],
    }),
    weekendDay: new ContainerBlock({
        width: new MatchParentSize({ weight: 1 }),
        height: new MatchParentSize(),
        items: [
            new ContainerBlock({
                width: new WrapContentSize(),
                height: new WrapContentSize(),
                orientation: 'horizontal',
                content_alignment_horizontal: 'left',
                items: [
                    new TextBlock({
                        ...text40m,
                        text: new Template('weekday'),
                        width: new WrapContentSize(),
                        height: new WrapContentSize(),
                    }),
                    new TextBlock({
                        ...text40m,
                        text_color: colorWhiteOpacity50,
                        text: new Template('date'),
                        width: new WrapContentSize(),
                        height: new WrapContentSize(),
                    }),
                ],
            }),
            new ContainerBlock({
                width: new WrapContentSize(),
                height: new WrapContentSize(),
                margins: {
                    top: 24,
                },
                orientation: 'horizontal',
                content_alignment_vertical: 'bottom',
                items: [
                    new ImageBlock({
                        image_url: new Template('icon'),
                        width: new FixedSize({ value: 132 }),
                        height: new FixedSize({ value: 132 }),
                    }),
                    new TextBlock({
                        ...title90m,
                        text: new Template('temp_day'),
                        width: new WrapContentSize(),
                        height: new WrapContentSize(),
                    }),
                    new TextBlock({
                        ...title64m,
                        text_color: colorWhiteOpacity50,
                        text: new Template('temp_night'),
                        margins: {
                            bottom: 8,
                        },
                        width: new WrapContentSize(),
                        height: new WrapContentSize(),
                    }),
                ],
            }),
        ],
    }),
};

export const tHelper = templateHelper(weatherTemplates);

type ContainerProps = {
    items: Div[];
    daypart: string;
    condition: string;
};
export const container = ({ items, daypart, condition }: ContainerProps) =>
    new ContainerBlock({
        width: new MatchParentSize(),
        height: new MatchParentSize(),
        background: [
            new ImageBackground({ image_url: getWeatherImage(daypart, condition) }),
        ],
        orientation: 'vertical',
        items,
    });

type NowBlockProps = {
    cityFormed: string;
    temp: number;
    iconType: string;
    feelsLike: number;
    title: string;
    when: string;
};
export const nowBlock = ({ cityFormed, temp, iconType, feelsLike, title, when }: NowBlockProps) =>
    new ContainerBlock({
        height: new WrapContentSize(),
        paddings: {
            left: 48,
            top: 48,
            right: 48,
        },
        items: [
            new TextBlock({
                ...text40m,
                text_color: colorWhiteOpacity50,
                width: new WrapContentSize(),
                height: new WrapContentSize(),
                text: when + ' ' + cityFormed,
            }),
            new ContainerBlock({
                orientation: 'horizontal',
                content_alignment_vertical: 'center',
                height: new WrapContentSize(),
                items: [
                    new TextBlock({
                        ...title146m,
                        width: new WrapContentSize(),
                        text: formatTemp(temp),
                    }),
                    new ImageBlock({
                        width: new FixedSize({ value: 176 }),
                        height: new FixedSize({ value: 176 }),
                        margins: {
                            top: 8,
                            left: 16,
                        },
                        image_url: formWeatherIconUrl(iconType ?? '', 176),
                    }),
                    new ContainerBlock({
                        width: new WrapContentSize(),
                        height: new WrapContentSize(),
                        orientation: 'vertical',
                        margins: {
                            left: 36,
                        },
                        items: [
                            new TextBlock({
                                ...text40m,
                                text_color: colorWhiteOpacity50,
                                width: new WrapContentSize(),
                                text: 'Ощущается как ' + formatTemp(feelsLike),
                            }),
                            new TextBlock({
                                ...text40m,
                                text_color: colorWhiteOpacity50,
                                width: new WrapContentSize(),
                                text: title,
                                margins: {
                                    top: 8,
                                },
                            }),
                        ],
                    }),
                ],
            }),
        ],
    });

type HoursGalleryProps = {
    hours: NAlice.NData.ITWeatherHourItem[];
};
export const hoursGallery = ({ hours }: HoursGalleryProps) =>
    new GalleryBlock({
        width: new MatchParentSize(),
        height: new WrapContentSize(),
        paddings: {
            left: 24,
            right: 48,
            top: 48,
        },
        orientation: 'horizontal',
        item_spacing: 24,
        items: hours.map(({ Hour, Icon, Temperature }) =>
            tHelper.weatherCard({
                time: formatHour(Hour ?? 0),
                icon: Icon ?? '',
                temp: formatTemp(Temperature ?? 0),
            }),
        ),
    });

export const dayParts = (parts: NAlice.NData.ITWeatherDayPartData[]) =>
    new GalleryBlock({
        orientation: 'horizontal',
        width: new WrapContentSize(),
        height: new WrapContentSize(),
        alignment_horizontal: 'center',
        paddings: {
            left: 48,
            right: 48,
        },
        item_spacing: 28,
        items: parts.map(({ DayPartType, Icon, Temperature }) =>
            tHelper.weatherCard({
                time: capitalize(dayPartsForms[DayPartType ?? 'day']),
                icon: Icon ?? '',
                temp: formatTemp(Temperature ?? 0),
            }),
        ),
    });

export const isWeekend = (
    days: Void<NAlice.NData.ITWeatherDayItem[]>,
): days is [NAlice.NData.ITWeatherDayItem, NAlice.NData.ITWeatherDayItem] =>
    days?.length === 2 && days[0].WeekDay === 6 && days[1].WeekDay === 0;

export const dateRange = (days: NAlice.NData.ITWeatherDayItem[]) => {
    switch (true) {
        case days.length === 7:
            return 'на неделю';
        case isWeekend(days):
            return 'на выходные';
    }

    return `на ${days.length} ${pluralize(days.length, { one: 'день', some: 'дня', many: 'дней' })}`;
};

export const daysGallery = (days: NAlice.NData.ITWeatherDayItem[]) =>
    new GalleryBlock({
        width: new MatchParentSize(),
        height: new WrapContentSize(),
        margins: {
            top: 80,
            bottom: 48,
        },
        paddings: {
            left: 48,
            right: 48,
        },
        orientation: 'horizontal',
        item_spacing: 24,
        items: days.map(({ Date: d, DayTemp, NightTemp, IconType }) =>
            tHelper.weatherDayCard({
                date: formatDate(new Date(d ?? 0), 'd MMMM'),
                weekDay: capitalize(formatDate(new Date(d ?? 0), 'cccc')),
                dayTemp: formatTemp(DayTemp ?? 0) + ' ',
                nightTemp: ' ' + formatTemp(NightTemp ?? 0),
                icon: formWeatherIconUrl(IconType ?? '', 176),
            }),
        ),
    });

export const weekendCast = (days: [NAlice.NData.ITWeatherDayItem, NAlice.NData.ITWeatherDayItem]) => {
    const divDays: Div[] = days.map(({ Date: VDate, IconType, DayTemp, NightTemp }) => tHelper.weekendDay({
        date: formatDate(new Date(VDate ?? 0), 'd MMMM'),
        weekday: capitalize(formatDate(new Date(VDate ?? 0), 'cccc')) + ' ',
        icon: formWeatherIconUrl(IconType ?? '', 132),
        temp_day: ' ' + formatTemp(DayTemp ?? 0) + ' ',
        temp_night: formatTemp(NightTemp ?? 0),
    }));
    divDays.splice(1, 0, new ContainerBlock({
        width: new FixedSize({ value: 1 }),
        height: new MatchParentSize(),
        margins: {
            left: 48,
            right: 48,
        },
        background: [
            new SolidBackground({ color: colorWhiteOpacity53 }),
        ],
        items: [new TextBlock({ text: ' ' })],
    }));

    return new ContainerBlock({
        width: new MatchParentSize(),
        height: new FixedSize({ value: 320 }),
        margins: {
            top: 170,
            bottom: 48,
        },
        paddings: {
            left: 48,
            right: 48,
        },
        orientation: 'horizontal',
        content_alignment_horizontal: 'center',
        items: divDays,
    });
};
