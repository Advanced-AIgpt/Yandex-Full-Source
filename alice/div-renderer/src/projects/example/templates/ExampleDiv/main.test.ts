import ExampleDiv from './ExampleDiv';
import { AnonymizeDataForSnapshot } from '../../../../common/helpers/dev';
import { createRequestState } from '../../../../registries/common';
import { MMRequest } from '../../../../common/helpers/MMRequest';

jest.mock('../../../../common/logger');

describe('example', () => {
    it('matches snapshot', () => {
        expect(AnonymizeDataForSnapshot(ExampleDiv({
            hello: 'world',
        }, new MMRequest({}, {}, {}), createRequestState()))).toMatchSnapshot();
    });
});
