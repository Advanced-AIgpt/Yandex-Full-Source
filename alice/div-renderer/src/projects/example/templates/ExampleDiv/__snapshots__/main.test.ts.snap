// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`example matches snapshot 1`] = `
Object {
  "card": Object {
    "log_id": "example_card",
    "states": Array [
      Object {
        "div": Object {
          "items": Array [
            Object {
              "hello": "world",
              "type": "greetings",
            },
          ],
          "type": "container",
        },
        "state_id": 0,
      },
    ],
    "variable_triggers": Array [],
    "variables": Array [],
  },
  "templates": Object {
    "greetings": Object {
      "$text": "hello",
      "text_color": "#ff0",
      "type": "text",
    },
  },
}
`;
